File: src/com/rian/difficultycalculator/beatmap/timings/EffectControlPoint.java
package com.rian.difficultycalculator.beatmap.timings;

/**
 * Represents a control point that applies an effect to a beatmap (e.g. kiai).
 */
public class EffectControlPoint extends ControlPoint {
    /**
     * Whether kiai is enabled at this control point.
     */
    public final boolean isKiai;

    /**
     * @param time   The time at which this control point takes effect, in milliseconds.
     * @param isKiai Whether kiai is enabled at this control point.
     */
    public EffectControlPoint(int time, boolean isKiai) {
        super(time);

        this.isKiai = isKiai;
    }

    @Override
    public boolean isRedundant(ControlPoint existing) {
        return existing instanceof EffectControlPoint && isKiai == ((EffectControlPoint) existing).isKiai;
    }
}

File: src/com/rian/osu/beatmap/timings/EffectControlPoint.kt
package com.rian.osu.beatmap.timings

/**
 * Represents a [ControlPoint] that applies an effect to a beatmap.
 */
class EffectControlPoint(
    /**
     * The time at which this [EffectControlPoint] takes effect, in milliseconds.
     */
    time: Double,

    /**
     * Whether kiai time is enabled at this [EffectControlPoint].
     */
    @JvmField
    val isKiai: Boolean
): ControlPoint(time) {
    override fun isRedundant(existing: ControlPoint) =
        existing is EffectControlPoint &&
        isKiai == existing.isKiai
}